### PR TITLE
Added kwargs to pass parameters to Compare Workspaces function.

### PR DESCRIPTION
**Pull request by dmitry-ganyushin**
### Description of work
Changed the method The method [mantid.testing.assert_almost_equal](https://github.com/mantidproject/mantid/blob/main/Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py#L11C5-L11C24)  to make it more useful for SNAPRed. The parameter rtol and etol got evaluated, the absence of both or presence of both generates an error. Additional parameters could be specified for the underlying mantid python function CompareWorkspaces

[EWM5931: [SNAPRed] assert_almost_equal improvements](https://ornlrse.clm.ibmcloud.com/ccm/resource/itemName/com.ibm.team.workitem.WorkItem/5931)

### CODE REVIEW

File: Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py [hunk 1]
```diff
@@ -29,19 +29,24 @@ def assert_almost_equal(Workspace1, Workspace2, rtol=0.0, atol=0.0):
...
-        tolerance = atol
 
-    if rtol:
-        tolerance = rtol
-        _rel = True
-
-    result, message = CompareWorkspaces(Workspace1, Workspace2, Tolerance=tolerance, ToleranceRelErr=_rel)
+    if rtol is not None and atol is not None:
+        raise ValueError("Specify rtol or atol, not both")
+    elif rtol in None and atol is None:
```
Comment (peterfpeterson) [inline Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py:39]:
```suggestion
    elif rtol is None and atol is None:
```

File: Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py [hunk 2]
```diff
@@ -29,19 +29,24 @@ def assert_almost_equal(Workspace1, Workspace2, rtol=0.0, atol=0.0):
...
     tolerance = 1e-10
-    if atol:
-        tolerance = atol
 
-    if rtol:
-        tolerance = rtol
-        _rel = True
-
-    result, message = CompareWorkspaces(Workspace1, Workspace2, Tolerance=tolerance, ToleranceRelErr=_rel)
+    if rtol is not None and atol is not None:
```
Comment (peterfpeterson) [inline Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py:37]: The story requested that if both were specified, run the algorithm twice, once with absolute tolerance, and once with relative tolerance.

File: Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py [hunk 3]
```diff
@@ -8,7 +8,7 @@
 from mantid.simpleapi import CompareWorkspaces
 
 
-def assert_almost_equal(Workspace1, Workspace2, rtol=0.0, atol=0.0):
+def assert_almost_equal(Workspace1, Workspace2, **kwargs):
```
Comment (peterfpeterson) [inline Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py:11]: The original design was that after you `from mantid.api.AlgorithmProperty import EMPTY_DBL` this should be
```suggestion
def assert_almost_equal(Workspace1, Workspace2, rtol=EMPTY_DBL, atol=EMPTY_DBL, **kwargs):
```
which requires more changes below. Alternatively, the signature could be
```
def assert_almost_equal(Workspace1, Workspace2, rtol=None, atol=None, **kwargs):
``` 
which would aid people in how to properly use this.

File: Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py [hunk 4]
```diff
@@ -8,7 +8,7 @@
...
     """
     Raises an assertion error if two workspaces are not within specified tolerance.
```
Comment (peterfpeterson) [inline Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py:14]: You should document the `kwargs` by linking to [CompareWorkspaces](https://docs.mantidproject.org/nightly/algorithms/CompareWorkspaces-v1.html) so people know what options they have.

File: Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py [hunk 5]
```diff
@@ -29,19 +29,24 @@ def assert_almost_equal(Workspace1, Workspace2, rtol=0.0, atol=0.0):
...
+    elif rtol in None and atol is None:
+        raise ValueError("Specify rtol or atol")
+    elif atol is not None:
+        if atol > 0.0:
+            tolerance = atol
+    elif rtol is not None:
+        if rtol > 0.0:
+            tolerance = rtol
+            _rel = True
+
```
Comment (peterfpeterson) [inline Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py:48]: Please error check that things aren't specified as normal arguments as well as kwargs. The biggies are `Workspace1`, `Workspace2`, `Tolerance`, and `ToleranceRelErr`.

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Please add tests and [release notes](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html) in addition to the comments made.

File: Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py (hunk 2 above)
Comment (dmitry-ganyushin) [reply Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py:37]: There was some contradiction between what is written in the story and the following discussion on slack channel. Ok, I am following requirements in the story.

File: Framework/PythonInterface/test/python/mantid/_testing/AssertAlmostEqualTest.py [hunk 6]
```diff
@@ -22,7 +22,7 @@ def setUpClass(self):
         self.ws3 = ws3
 
     def test_simple(self):
-        assert_almost_equal(self.ws1, self.ws1)
+        assert_almost_equal(self.ws1, self.ws2, atol=1, rtol=1)
```
Comment (peterfpeterson) [inline Framework/PythonInterface/test/python/mantid/_testing/AssertAlmostEqualTest.py:25]: I like that you test having both specified. Please add a test for specifying only `atol` and only `rtol`

Comment (dmitry-ganyushin) [reply Framework/PythonInterface/test/python/mantid/_testing/AssertAlmostEqualTest.py:25]: rtol and atol tests existed initially.

**Review by peterfpeterson — APPROVED** [review]

File: Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py [hunk 7]
```diff
@@ -26,23 +27,26 @@ def assert_almost_equal(Workspace1, Workspace2, rtol=0.0, atol=0.0):
...
-        msg = ", ".join(msg_dict["Message"]) + f", Workspaces are not within tolerance ({tolerance})"
-        raise AssertionError(msg)
+    # check arguments
+    if len(set(kwargs.keys()).intersection({"Workspace1", "Workspace2", "Tolerance", "ToleranceRelErr"})):
+        raise ValueError("Workspace1, Workspace2, Tolerance, ToleranceRelErr cannot be passed as additional parameters")
+
+    if rtol == Property.EMPTY_DBL and atol == Property.EMPTY_DBL:
+        raise ValueError("Specify rtol or atol")
+
+    if rtol > Property.EMPTY_DBL:
```
Comment (rboston628) [inline Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py:40]: `Property.EMPTY_DBL` is [equivalent to `DBL_MAX` in C++](https://github.com/mantidproject/mantid/blob/65408482e579ec532b373a45032b5ca40a45ec52/Framework/Kernel/inc/MantidKernel/EmptyValues.h#L43).  This condition will _never_ be true.  Likewise for `atol` below.  The result is that this test fixture will _always_ pass, even if the workspaces are not equal.

File: Framework/PythonInterface/test/python/mantid/_testing/AssertAlmostEqualTest.py [hunk 8]
```diff
@@ -33,11 +33,8 @@ def test_rtol(self):
         assert_almost_equal(self.ws1, self.ws3, rtol=0.7)
 
     def test_raises(self):
-        with self.assertRaises(AssertionError):
-            assert_almost_equal(self.ws1, self.ws2)
```
Comment (rboston628) [inline Framework/PythonInterface/test/python/mantid/_testing/AssertAlmostEqualTest.py:37]: This negative test that was removed would have prevented the error above.

File: Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py (hunk 7 above)
Comment (peterfpeterson) [reply Framework/PythonInterface/mantid/_testing/AssertAlmostEqual.py:40]: `Property.EMPTY_DBL` is a special value that optional float properties can be set to. For interaction with `CompareWorkspaces`, both `atol` and `rtol` are (separately) passed in as [Tolerance parameter](https://github.com/mantidproject/mantid/blob/main/Framework/Algorithms/src/CompareWorkspaces.cpp#L123) and read by the algorithm [here](https://github.com/mantidproject/mantid/blob/main/Framework/Algorithms/src/CompareWorkspaces.cpp#L123). This line should probably be changed to verify that the value isn't set to `EMPTY_DBL`.

File: Framework/PythonInterface/test/python/mantid/_testing/AssertAlmostEqualTest.py (hunk 8 above)
Comment (peterfpeterson) [reply Framework/PythonInterface/test/python/mantid/_testing/AssertAlmostEqualTest.py:37]: You are being mislead by the diff (or simply misreading it). The original code was
```python
        with self.assertRaises(AssertionError):
            assert_almost_equal(self.ws1, self.ws2)

        with self.assertRaises(ValueError):
            assert_almost_equal(self.ws1, self.ws2, atol=1, rtol=1)
```
and the new version is
```python
        with self.assertRaises(AssertionError):
            assert_almost_equal(self.ws1, self.ws2)
```
not specifying either is still an error. Specifying both is ok.

Comment (rboston628) [reply Framework/PythonInterface/test/python/mantid/_testing/AssertAlmostEqualTest.py:37]: Here is the entire test suite:
``` python
class AssertAlmostEqualTest(unittest.TestCase):
    @classmethod
    def setUpClass(self):
        ws1 = CreateWorkspace(DataX=[0, 1, 2, 3, 4, 5], DataY=[1, 1, 1, 1, 1, 1])
        ws2 = CreateWorkspace(DataX=[0.09, 1.09, 2.09, 3.09, 4.09, 5.09], DataY=[1, 1, 1, 1, 1, 1])
        ws3 = CreateWorkspace(DataX=[0, 2, 4, 6, 8, 10], DataY=[1, 1, 1, 1, 1, 1])
        self.ws1 = ws1
        self.ws2 = ws2
        self.ws3 = ws3

    def test_simple(self):
        assert_almost_equal(self.ws1, self.ws2, atol=1, rtol=1)

    def test_atol(self):
        # compare (ws1 - ws2) < atol
        assert_almost_equal(self.ws1, self.ws2, atol=0.1)

    def test_rtol(self):
        # compare (ws1 - ws2) / (0.5 * (ws1 + ws2)) < rtol
        assert_almost_equal(self.ws1, self.ws3, rtol=0.7)

    def test_raises(self):
        with self.assertRaises(ValueError):
            assert_almost_equal(self.ws1, self.ws2)
```
The only "negative" test remaining is asserting that a value error is raised, which happens if neither `atol` nor `rtol` are specified.

Comment (rboston628) [reply Framework/PythonInterface/test/python/mantid/_testing/AssertAlmostEqualTest.py:37]: Ran the following in a version of workbench containing this change, which passed
``` python
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

from mantid.testing import assert_almost_equal

ws1 = CreateSingleValuedWorkspace(DataValue=1.0)
ws2 = CreateSingleValuedWorkspace(DataValue=2.0)
assert_almost_equal(ws1, ws2, atol=0.0, rtol=0.0)
print("assert passes")
```

Comment (dmitry-ganyushin) [reply Framework/PythonInterface/test/python/mantid/_testing/AssertAlmostEqualTest.py:37]: Should we just treat all `atol` and `rtol` which are less than  `Property.EMPTY_DBL` equal to `Property.EMPTY_DBL` ? We can catch this special corner case then.

Comment (peterfpeterson) [reply Framework/PythonInterface/test/python/mantid/_testing/AssertAlmostEqualTest.py:37]: `EMPTY_DBL` is a special value. One should only ever check for equality
